src/simulators/matrix_product_state/matrix_product_state_internal.cpp

```diff
@@ -520,14 +520,81 @@ void MPS::apply_matrix(const reg_t & qubits, const cmatrix_t &mat)
   case 1: 
     q_reg_[qubits[0]].apply_matrix(mat);
     break;
-  case 2:
-    apply_2_qubit_gate(qubits[0], qubits[1], su4, mat);
-    break;
+    //  case 2:
+    //    apply_2_qubit_gate(qubits[0], qubits[1], su4, mat);
+    //    break;
   default:
-    throw std::invalid_argument("currently support apply_matrix for 1 or 2 qubits only");
+    apply_multi_qubit_gate(qubits, mat);
   }
 }
 
+void MPS::apply_multi_qubit_gate(const reg_t &qubits,
+				 const cmatrix_t &mat) {
+  uint_t num_qubits = qubits.size();
+  // need to reverse qubits because that is the way they
+  // are defined in the Qiskit interface
+  reg_t reversed_qubits = qubits;
+  std::reverse(reversed_qubits.begin(), reversed_qubits.end()); 
+
+  bool ordered = true;
+  for (uint_t index=0; index < qubits.size()-1; index++) {
+    if (reversed_qubits[index]+1 != reversed_qubits[index+1]){
+      ordered = false;
+      break;
+    }
+  }
+  reg_t actual_indices(num_qubits_);
+  std::iota( std::begin(actual_indices), std::end(actual_indices), 0);
+  reg_t target_qubits(num_qubits);
+  if (ordered) {
+    target_qubits = reversed_qubits;
+  } else {
+    move_qubits_to_right_end(reversed_qubits, target_qubits, actual_indices);
+  }
+
+  uint_t first = target_qubits.front();
+  MPS_Tensor sub_tensor(state_vec_as_MPS(first, first+num_qubits-1));
+
+  sub_tensor.apply_matrix(mat, false /* swapped */);
```

Comment:
The fact that `apply_matrix` needs to swap the contained matrix before computing the tensor, by passing a `true/false` value is kind of confusing. In order to preserve `Single Responsibility` principle I would definitely create a `swap` method in `MPS_Tensor` and express this like:
```c++
sub_tensor.swap();
sub_tensor.apply_matrix(mat);
```